opt: add indexed var for inverted column when execbuilding inverted joins

rytaft opened the pull request:
This commit fixes an issue identified in #50709 in which the `execbuilder`
code for building inverted joins was failing due to the lack of an indexed
var for the column indexed by the inverted index. The indexed var is
necessary since the inverted expression contains a reference to the inverted
column. This commit fixes the issue by including the column in the indexed var
helper.

This commit does not include a release note since this issue was introduced
a few days ago and has not been released.

Release note: None

rytaft committed:
opt: add indexed var for inverted column when execbuilding inverted joins

This commit fixes an issue identified in #50709 in which the execbuilder
code for building inverted joins was failing due to the lack of an indexed
var for the column indexed by the inverted index. The indexed var is
necessary since the inverted expression contains a reference to the inverted
column. This commit fixes the issue by including the column in the indexed var
helper.

This commit does not include a release note since this issue was introduced
a few days ago and has not been released.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50754)
<!-- Reviewable:end -->

rytaft commented:
Hold off on reviewing this for now -- I'm testing it with #50709 to see if I can identify any other issues.





<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50754)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @sumeerbhola)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50754#-MB-FQQ80T8Lq3t8Hw--:-MB-FQQ80T8Lq3t8Hw-0:big3hxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0a4fe7559159d036dc16e3def4568b84832fe0f/pkg/sql/opt/exec/execbuilder/relational.go#L1403)):*
> ```Go
> 
> 	lookupOrdinals, lookupColMap := b.getColumns(lookupCols, join.Table)
> 	allCols := joinOutputMap(input.outputCols, lookupColMap)
> ```

I think this will include this inverted column in `allCols` which is the output of this join, which isn't quite correct. 

Are we going down the path of the `invertedFilterer` integration where we couldn't figure out how to do the col mappings in a way that the inverted column was provided in the input but not the output and needed the projection?


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50754#-MB-FQQ80T8Lq3t8Hw--:-MB-Of8E0e3vDXr859gz:b-4v7zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0a4fe7559159d036dc16e3def4568b84832fe0f/pkg/sql/opt/exec/execbuilder/relational.go#L1403)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think this will include this inverted column in `allCols` which is the output of this join, which isn't quite correct. 

Are we going down the path of the `invertedFilterer` integration where we couldn't figure out how to do the col mappings in a way that the inverted column was provided in the input but not the output and needed the projection?
</blockquote></details>

Yea, just added the post projection here. PTAL.

I tried to remove it from the outputCols directly, but it still showed up in the EXPLAIN, so I don't think it was working.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50754#-MB-FQQ80T8Lq3t8Hw--:-MB-Yp0l9xb07L5f7kWl:b-os7hzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0a4fe7559159d036dc16e3def4568b84832fe0f/pkg/sql/opt/exec/execbuilder/relational.go#L1403)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Yea, just added the post projection here. PTAL.

I tried to remove it from the outputCols directly, but it still showed up in the EXPLAIN, so I don't think it was working.
</blockquote></details>

cc @raduberinde do you know if there is a better way to do this?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50754)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50754#-MB-FQQ80T8Lq3t8Hw--:-MB0OUpJ28rXYUAbV-u9:b-a9zrx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0a4fe7559159d036dc16e3def4568b84832fe0f/pkg/sql/opt/exec/execbuilder/relational.go#L1403)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

cc @raduberinde do you know if there is a better way to do this?
</blockquote></details>

Not sure I understand the issue.. `ConstructInvertedJoin` is documented as producing all the lookup columns, if that's the case, yes a projection is needed.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1403 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50754#-MB-FQQ80T8Lq3t8Hw--:-MB0R8g_4U6LrQPQRhGM:b-3xffca) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0a4fe7559159d036dc16e3def4568b84832fe0f/pkg/sql/opt/exec/execbuilder/relational.go#L1403)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not sure I understand the issue.. `ConstructInvertedJoin` is documented as producing all the lookup columns, if that's the case, yes a projection is needed.
</blockquote></details>

The issue is that it should be possible for the `invertedJoiner` DistSQL processor to avoid returning the inverted column, which would make the simple project that I added at the end of this function unnecessary. But I couldn't figure out how to pass the correct columns (i.e., get the expected EXPLAIN output and avoid errors and panics) without the project. I'm not sure how much overhead the project actually adds, so this may not be a real issue, but just wanted to get your opinion.

The documentation for `ConstructInvertedJoin` was stale, so I just updated it to change the sentence "Note that lookupCols does not include the inverted column" to  "Note that lookupCols includes the inverted column."


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 4 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50754#-:-MB1fVQ9BA3Mz7evXV73:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+





<!-- Sent from Reviewable.io -->

rytaft commented:
Opened https://github.com/cockroachdb/cockroach/issues/50795

bors r+





<!-- Sent from Reviewable.io -->
